src/backtrace_codec.ml

```diff
@@ -173,9 +174,19 @@ module Reader = struct
         let pred' = cache_pred.(pred) in
         let bbuf = put_bbuf bbuf pos cache_loc.(pred') in
         predict pred' bbuf (pos + 1) i (n-1) in
-    let (bbuf, pos) = decode 0 cache.last_backtrace common_pfx_len nencoded in
-    cache.last_backtrace <- bbuf;
-    (bbuf, pos)
+    if common_pfx_len <= cache.last_backtrace_len then begin
+      let (bbuf, pos) = decode 0 cache.last_backtrace common_pfx_len nencoded in
+      cache.last_backtrace <- bbuf;
+      cache.last_backtrace_len <- pos;
+      (bbuf, pos)
+    end else begin
+      (* This can occur if the last backtrace was truncated, and the current
+         backtrace shares a long prefix with it. Return the amount of backtrace
+         that we have. (We still go through the motions of decoding to ensure
+         that the location cache is updated correctly) *)
+      let (_bbuf, _pos) = decode 0 [| |] 0 nencoded in
+      (cache.last_backtrace, cache.last_backtrace_len)
```

Comment:
Just to check my understanding, the rationale here is that the last backtrace was so big it got truncated to some maximum length, and this backtrace has a common prefix that's longer than that length, so it effectively gets truncated to exactly the same backtrace?

Reply:
Yeah, I think that's the only reasonable thing we can do here. The actual backtrace is:

    [pppppp???????xxxxxx]

where `pppppp???????` is the common prefix with the last one, `pppppp` is the untruncated part that we have, `???????` is the bit we lost to truncation, and `xxxxxx` is the new suffix that's been encoded here. The longest prefix that we can correctly produce is just `pppppp`, because we have no idea what's between that and `xxxxxx`

Reply:
Oh, I did not realize `xxxxxx` was actually available.
I guess in the general case `pppppp???????xxxxxx????` is what we have (`xxxxxx` can also be truncated).